fontir/src/variations.rs

```diff
@@ -113,6 +114,98 @@ impl VariationModel {
     pub fn default_location(&self) -> &NormalizedLocation {
         &self.default
     }
+
+    /// Convert absolute positions at master locations to offsets.
+    ///
+    /// <ul>
+    ///     <li>All keys in points must be known to the variation model.</li>
+    ///     <li>All point vectors must have the same length.</li>
+    ///     <li>A point vector for the default location is required</li>
+    /// </ul>
+    ///
+    /// Note that it is NOT required to provide a point sequence for every location known
+    /// to the variation model.
+    ///
+    /// P is the point type, meant to be absolute position in 1 or 2 dimensional space.
+    /// V is the vector type.
+    ///
+    /// For 2d [kurbo::Point] and [kurbo::Vec2] would be typical choices. For 1d a floating
+    /// point primitive should suffice.
+    ///
+    /// Returns a delta, as the vector type, for every input point. Intended use is to support
+    /// construction of a variation store.
+    ///
+    /// Rust version of <https://github.com/fonttools/fonttools/blob/3b9a73ff8379ab49d3ce35aaaaf04b3a7d9d1655/Lib/fontTools/varLib/models.py#L449-L461>
+    pub fn deltas<P, V>(
+        &self,
+        point_seqs: &HashMap<NormalizedLocation, Vec<P>>,
+    ) -> Result<HashMap<NormalizedLocation, Vec<V>>, DeltaError>
+    where
+        P: Copy + Default + Sub<P, Output = V>,
+        V: Copy + Mul<f64, Output = V> + Sub<V, Output = V>,
+    {
+        let Some(defaults) = point_seqs.get(&self.default) else {
+            return Err(DeltaError::DefaultUndefined);
+        };
+        if point_seqs.values().any(|pts| pts.len() != defaults.len()) {
+            return Err(DeltaError::InconsistentNumbersOfPoints);
+        }
+
+        let mut result: HashMap<NormalizedLocation, Vec<V>> = HashMap::new();
+        // self.locations is sorted such that[i] is only influenced by[i+1..N]
+        // so we know subsequent spins won't invalidate our delta if we go in the same order
+        for (loc, points, master_influences) in
+            self.locations
+                .iter()
+                .enumerate()
+                .filter_map(|(loc_idx, loc)| {
+                    point_seqs
+                        .get(loc)
+                        .map(|points| (loc, points, &self.delta_weights[loc_idx]))
+                })
+        {
+            let mut deltas = Vec::new();
+            for (idx, point) in points.iter().enumerate() {
+                let initial_vector: V = *point - Default::default();
```

Comment:
I'm staring at this like an idiot: it feels like a convoluted way to convert from the input type to the vector type? Am I missing anything?

Reply:
That's correct. There isn't a nice trait that works for the types - I was hoping for V: From<P> or some such - we want to use this for so the other option I came up with was to pass in a function that converts P => V. I think this is cleaner than that.

@dfrg wdyt about Vec2 having From<Point>, possibly also vice versa?